### DISCUSSION TITLE
MTU_MIN, MTU_MAX

### DISCUSSION THREAD

**Discussion by jvfranklin** [rsmarples/dhcpcd#344] [General]
In src/dhcp.c make_message(), there's this code enforcing a minimum MTU:

else if (mtu < MTU_MIN) {
		if (if_setmtu(ifp, MTU_MIN) == -1)

MTU_MIN is defined in src/dhcp-common.h:
#define	IP_UDP_SIZE		  28
#define	MTU_MAX			1500 - IP_UDP_SIZE
#define	MTU_MIN			 576 + IP_UDP_SIZE

I work on an embedded Linux system that allows a minimum MTU on the link of 576. It appears from my initial testing that dhcpcd is able to send and receive DHCP messages without any issue if I change MTU_MIN=576. Our previous DHCP client was also able to handle this, as well. I don't see any fragmentation in the packet captures. The ethernet frames are less than 400 bytes, at least for my environment.

Is there a history on this code that requires MTU_MIN to be 604?

Also wondering about the MTU_MAX, though I haven't explored its effect in our product yet.

**Comment by rsmarples** [comment]
https://datatracker.ietf.org/doc/html/rfc2131#section-2

> The 'options' field is now variable length. A DHCP client must be
   prepared to receive DHCP messages with an 'options' field of at least
   length 312 octets.  This requirement implies that a DHCP client must
   be prepared to receive a message of up to 576 octets, the minimum IP
   datagram size an IP host must be prepared to accept.

https://datatracker.ietf.org/doc/html/rfc1542#section-2.1

> The IP Total Length and UDP Length must be large enough to
        contain the minimal BOOTP header of 300 octets (in the UDP
        data field) specified in [[1](https://datatracker.ietf.org/doc/html/rfc1542#ref-1)].

From my testing over the years, the BOOTP message size of 300 octets is a real thing. Some servers reject a BOOTP message that's smaller.
Because of the similar language, my assumption was 576 was the minimum DHCP message size, NOT including the IP and UDP headers.

However, testing shows my assumption is not correct. 576 - (IP + UDP header size of 28) == bootp message size with vendor length of 312.
I will correct this.

**Comment by rsmarples** [comment]
We'll fix this in #345

**Comment by rsmarples** [comment]
@jvfranklin I can't seem to tag you in that issue. Anyway, could you review the PR please and make sure it fixes things for you?

**Reply by jvfranklin** [reply]
Thanks for the quick turnaround! This works great.
